kubernetes/base/deployment.yaml

```diff
@@ -1,22 +1,85 @@
 apiVersion: apps/v1
 kind: Deployment
+
 metadata:
   name: chat-service
   labels:
     app: chat-service
+
 spec:
   replicas: 1
+
   selector:
     matchLabels:
       app: chat-service
+
   template:
     metadata:
       labels:
         app: chat-service
+
     spec:
+      serviceAccountName: chat-service
+
+      volumes:
+      # Configuration files for the ejabberd process
+      - name: ejabberd-config
+        configMap:
+          name: ejabberd-config
+      # Secret configuration files for the ejabberd process
+      - name: ejabberd-config-secret
+        secret:
+          secretName: ejabberd-config-secret
+      # Configuration files for the vault and consul-template agents
+      - name: vault-config
+        configMap: ejabberd-vault-config
+        items:
+          - key: vault-agent-config.hcl
+            path: vault-agent-config.hcl
+          - key: consul-template-config.hcl
+            path: consul-template-config.hcl
+      # Stores the vault authentication token in RAM-backed tmpfs
+      - name: vault-token
+        emptyDir:
+          medium: Memory
+      # Stores files containing vault secrets rendered by consul-template
+      - name: vault-secrets
+        emptyDir: {}
+
+      initContainers:
+      # Vault agent manages authentication with vault on pod startup
+      - name: vault-agent-auth
+        image: vault:1.2.3 # Latest as of 2019-10-10
+        volumeMounts:
+        - name: vault-config
+          mountPath: /etc/vault
+        - name: vault-token
+          mountPath: /home/vault
+        env:
+        - name: VAULT_ADDR
+          value: http://vault.default:8200 # Vault is currently deployed in the default namespace
+        args: [ "agent", "-config=/etc/vault/vault-agent-config.hcl" ]
+
       containers:
-      - image: docker.dev.skillz.com:5000/ejabberd:latest
-        name: chat-service
+      # consul-template connects to vault and populates templates with the values of secrets
```

Comment:
Everything that wants to use secrets needs this sidecar? That's a bummer.

Reply:
No this is just one possible implementation that worked for me without modifying the ejabberd source.
There are plenty of other ways to get secrets from vault and I would actually encourage connecting to vault directly from the application instead of rendering templates so secrets are always kept in memory:
https://www.vaultproject.io/api/libraries.html
https://spring.io/guides/gs/vault-config/